// @antv/g2plot/lib/plots/tiny-area/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TinyArea = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var constants_1 = require("../tiny-line/constants");
var adaptor_1 = require("./adaptor");
var TinyArea = /** @class */ (function (_super) {
    tslib_1.__extends(TinyArea, _super);
    function TinyArea() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'tiny-area';
        return _this;
    }
    TinyArea.prototype.getDefaultOptions = function () {
        return {
            appendPadding: 2,
            tooltip: tslib_1.__assign({}, constants_1.DEFAULT_TOOLTIP_OPTIONS),
            // 默认样式
            color: 'l(90) 0:#E5EDFE 1:#ffffff',
            areaStyle: {
                fillOpacity: 0.6,
            },
            line: {
                size: 1,
                color: '#5B8FF9',
            },
        };
    };
    /**
     * 获取 迷你面积图 的适配器
     */
    TinyArea.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return TinyArea;
}(plot_1.Plot));
exports.TinyArea = TinyArea;
