// @antv/g2plot/lib/plots/tiny-column/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = void 0;
var common_1 = require("../../adaptor/common");
var utils_1 = require("../../utils");
var geometries_1 = require("../../adaptor/geometries");
var constants_1 = require("../tiny-line/constants");
var utils_2 = require("../tiny-line/utils");
var data_1 = require("../../utils/data");
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var _a, _b;
    var chart = params.chart, options = params.options;
    var data = options.data, xAxis = options.xAxis, yAxis = options.yAxis, color = options.color, columnStyle = options.columnStyle, columnWidthRatio = options.columnWidthRatio;
    var seriesData = utils_2.getTinyData(data);
    chart.data(seriesData);
    var p = utils_1.deepAssign({}, params, {
        options: {
            xField: constants_1.X_FIELD,
            yField: constants_1.Y_FIELD,
            widthRatio: columnWidthRatio,
            interval: {
                style: columnStyle,
                color: color,
            },
        },
    });
    geometries_1.interval(p);
    chart.axis(false);
    chart.legend(false);
    chart.interaction('element-active');
    // scale
    common_1.scale((_a = {},
        _a[constants_1.X_FIELD] = xAxis,
        _a[constants_1.Y_FIELD] = yAxis,
        _a), (_b = {},
        _b[constants_1.X_FIELD] = {
            type: 'cat',
        },
        _b[constants_1.Y_FIELD] = data_1.adjustYMetaByZero(seriesData, constants_1.Y_FIELD),
        _b))(params);
    return params;
}
/**
 * 迷你柱形图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    return utils_1.flow(geometry, common_1.tooltip, common_1.theme, common_1.animation, common_1.annotation())(params);
}
exports.adaptor = adaptor;
