// @antv/g2plot/lib/plots/tiny-column/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TinyColumn = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var utils_1 = require("../tiny-line/utils");
var adaptor_1 = require("./adaptor");
var constants_1 = require("./constants");
var TinyColumn = /** @class */ (function (_super) {
    tslib_1.__extends(TinyColumn, _super);
    function TinyColumn() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'tiny-column';
        return _this;
    }
    /**
     * @override
     * @param data
     */
    TinyColumn.prototype.changeData = function (data) {
        this.updateOption({ data: data });
        this.chart.changeData(utils_1.getTinyData(data));
    };
    TinyColumn.prototype.getDefaultOptions = function () {
        return {
            appendPadding: 2,
            tooltip: tslib_1.__assign({}, constants_1.DEFAULT_TOOLTIP_OPTIONS),
            animation: true,
        };
    };
    /**
     * 获取 迷你柱形图 的适配器
     */
    TinyColumn.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return TinyColumn;
}(plot_1.Plot));
exports.TinyColumn = TinyColumn;
