// @antv/g2plot/esm/plots/scatter/index.js
import { __extends } from "tslib";
import { Plot } from '../../core/plot';
import { deepAssign } from '../../utils';
import { adaptor } from './adaptor';
import './interaction';
var Scatter = /** @class */ (function (_super) {
    __extends(Scatter, _super);
    function Scatter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'point';
        return _this;
    }
    /**
     * 获取散点图的适配器
     */
    Scatter.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    Scatter.prototype.getDefaultOptions = function () {
        return deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            size: 4,
            tooltip: {
                showTitle: false,
                showMarkers: false,
                showCrosshairs: true,
                crosshairs: {
                    type: 'xy',
                },
            },
        });
    };
    return Scatter;
}(Plot));
export { Scatter };
