// @antv/g2plot/lib/plots/tiny-line/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = void 0;
var utils_1 = require("../../utils");
var common_1 = require("../../adaptor/common");
var geometries_1 = require("../../adaptor/geometries");
var data_1 = require("../../utils/data");
var constants_1 = require("./constants");
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var _a, _b;
    var chart = params.chart, options = params.options;
    var data = options.data, xAxis = options.xAxis, yAxis = options.yAxis, color = options.color, lineStyle = options.lineStyle, pointMapping = options.point;
    var seriesData = data.map(function (y, x) {
        return { x: x, y: y };
    });
    chart.data(seriesData);
    // line geometry 处理
    var p = utils_1.deepAssign({}, params, {
        options: {
            xField: constants_1.X_FIELD,
            yField: constants_1.Y_FIELD,
            line: {
                color: color,
                style: lineStyle,
            },
            point: pointMapping,
        },
    });
    geometries_1.line(p);
    geometries_1.point(p);
    chart.axis(false);
    chart.legend(false);
    // scale
    common_1.scale((_a = {},
        _a[constants_1.X_FIELD] = xAxis,
        _a[constants_1.Y_FIELD] = yAxis,
        _a), (_b = {},
        _b[constants_1.X_FIELD] = {
            type: 'cat',
        },
        _b[constants_1.Y_FIELD] = data_1.adjustYMetaByZero(seriesData, constants_1.Y_FIELD),
        _b))(params);
    return params;
}
/**
 * 迷你折线图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    return utils_1.flow(geometry, common_1.theme, common_1.tooltip, common_1.animation, common_1.annotation())(params);
}
exports.adaptor = adaptor;
