// @antv/g2plot/esm/plots/bullet/index.js
import { __extends } from "tslib";
import { Plot } from '../../core/plot';
import { deepAssign } from '../../utils';
import { adaptor } from './adaptor';
var Bullet = /** @class */ (function (_super) {
    __extends(Bullet, _super);
    function Bullet() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'bullet';
        return _this;
    }
    /**
     * 获取子弹图的适配器
     */
    Bullet.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    Bullet.prototype.getDefaultOptions = function () {
        return deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            layout: 'horizontal',
            size: {
                range: 30,
                measure: 20,
                target: 20,
            },
            xAxis: {
                tickLine: false,
            },
            bulletStyle: {
                range: {
                    fillOpacity: 0.5,
                },
            },
            label: {
                measure: {
                    position: 'right',
                },
            },
        });
    };
    return Bullet;
}(Plot));
export { Bullet };
