// @antv/g2plot/lib/plots/liquid/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Liquid = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
// register liquid shape
require("./shapes/liquid");
/**
 * 传说中的水波图
 */
var Liquid = /** @class */ (function (_super) {
    tslib_1.__extends(Liquid, _super);
    function Liquid() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'liquid';
        return _this;
    }
    Liquid.prototype.getDefaultOptions = function () {
        return {
            color: '#6a99f9',
            radius: 0.9,
            statistic: {
                title: false,
                content: {
                    formatter: function (_a) {
                        var percent = _a.percent;
                        return (percent * 100).toFixed(2) + "%";
                    },
                    style: {
                        opacity: 0.75,
                        fontSize: '30px',
                        textAlign: 'center',
                    },
                },
            },
        };
    };
    /**
     * 更新数据
     * @param percent
     */
    Liquid.prototype.changeData = function (percent) {
        this.update({
            percent: percent,
        });
    };
    /**
     * 获取适配器
     */
    Liquid.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Liquid;
}(plot_1.Plot));
exports.Liquid = Liquid;
