// @antv/g2plot/lib/plots/histogram/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Histogram = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var utils_1 = require("../../utils");
var adaptor_1 = require("./adaptor");
var Histogram = /** @class */ (function (_super) {
    tslib_1.__extends(Histogram, _super);
    function Histogram() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'histogram';
        return _this;
    }
    Histogram.prototype.getDefaultOptions = function () {
        return utils_1.deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            // @ts-ignore
            columnStyle: {
                stroke: '#FFFFFF',
            },
            tooltip: {
                shared: true,
                showMarkers: false,
            },
            interactions: [{ type: 'active-region' }],
        });
    };
    /**
     * 获取直方图的适配器
     */
    Histogram.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Histogram;
}(plot_1.Plot));
exports.Histogram = Histogram;
