// @ant-design/charts/es/graph/dagreFundFlow.js
import React, { useEffect } from 'react';
import G6 from '@antv/g6/es';
import { ErrorBoundary } from '../base';
import { processMinimap, getGraphSize } from './util';
import useGraph from '../hooks/useGraph';
var defaultStateStyles = {
    hover: {
        stroke: '#1890ff',
        lineWidth: 2
    }
};
var defaultNodeSize = [150, 30];
var defaultNodeStyle = {
    stroke: '#72CC4A',
    fill: '#f00'
};
var defaultNodeAnchorPoints = [[0.5, 0], [0.5, 1]];
var defaultEdgeStyle = {
    stroke: '#91d5ff',
    endArrow: {
        path: G6.Arrow.vee(10, 10),
        fill: '#91d5ff',
    }
};
var defaultLayout = {
    type: 'dagre',
    rankdir: 'LR',
    nodesep: 30,
    ranksep: 50,
};
var defaultLabelCfg = {
    style: {
        fill: '#000000A6',
        fontSize: 10,
    }
};
var graph;
var DagreFundFlowGraph = function (_a) {
    var data = _a.data, className = _a.className, style = _a.style, width = _a.width, height = _a.height, _b = _a.nodeType, nodeType = _b === void 0 ? 'round-rect' : _b, _c = _a.edgeType, edgeType = _c === void 0 ? 'fund-polyline' : _c, _d = _a.behaviors, behaviors = _d === void 0 ? ['zoom-canvas', 'drag-canvas'] : _d, _e = _a.nodeSize, nodeSize = _e === void 0 ? defaultNodeSize : _e, _f = _a.nodeLabelCfg, nodeLabelCfg = _f === void 0 ? defaultLabelCfg : _f, _g = _a.edgeLabelCfg, edgeLabelCfg = _g === void 0 ? defaultLabelCfg : _g, _h = _a.nodeAnchorPoints, nodeAnchorPoints = _h === void 0 ? defaultNodeAnchorPoints : _h, _j = _a.layout, layout = _j === void 0 ? defaultLayout : _j, minimapCfg = _a.minimapCfg, _k = _a.nodeStyle, nodeStyle = _k === void 0 ? defaultNodeStyle : _k, _l = _a.edgeStyle, edgeStyle = _l === void 0 ? defaultEdgeStyle : _l, _m = _a.nodeStateStyles, nodeStateStyles = _m === void 0 ? defaultStateStyles : _m, _o = _a.edgeStateStyles, edgeStateStyles = _o === void 0 ? defaultStateStyles : _o, _p = _a.colorMap, colorMap = _p === void 0 ? {} : _p, handleEdgeClick = _a.handleEdgeClick, handleEdgeHover = _a.handleEdgeHover, handleEdgeUnHover = _a.handleEdgeUnHover, handleNodeClick = _a.handleNodeClick, handleNodeHover = _a.handleNodeHover, handleNodeUnHover = _a.handleNodeUnHover, handleCanvasClick = _a.handleCanvasClick, graphRef = _a.graphRef;
    var container = React.useRef(null);
    var props = {
        data: data,
        className: className,
        style: style,
        width: width,
        height: height,
        nodeType: nodeType,
        edgeType: edgeType,
        behaviors: behaviors,
        nodeSize: nodeSize,
        nodeLabelCfg: nodeLabelCfg,
        edgeLabelCfg: edgeLabelCfg,
        nodeAnchorPoints: nodeAnchorPoints,
        layout: layout,
        minimapCfg: minimapCfg,
        nodeStyle: nodeStyle,
        edgeStyle: edgeStyle,
        nodeStateStyles: nodeStateStyles,
        edgeStateStyles: edgeStateStyles,
        colorMap: colorMap,
        handleEdgeClick: handleEdgeClick,
        handleEdgeHover: handleEdgeHover,
        handleEdgeUnHover: handleEdgeUnHover,
        handleNodeClick: handleNodeClick,
        handleNodeHover: handleNodeHover,
        handleNodeUnHover: handleNodeUnHover,
        handleCanvasClick: handleCanvasClick,
        graphRef: graphRef
    };
    useGraph(graph, props, container);
    useEffect(function () {
        var graphSize = getGraphSize(width, height, container);
        if (!graph || graph.destroyed) {
            graph = new G6.Graph({
                container: container.current,
                width: graphSize[0],
                height: graphSize[1],
                modes: {
                    default: behaviors,
                },
                defaultNode: {
                    type: nodeType,
                    size: nodeSize,
                    style: nodeStyle,
                    anchorPoints: nodeAnchorPoints,
                    labelCfg: nodeLabelCfg,
                },
                defaultEdge: {
                    type: edgeType,
                    style: edgeStyle,
                    colorMap: colorMap,
                    labelCfg: edgeLabelCfg,
                },
                nodeStateStyles: nodeStateStyles,
                edgeStateStyles: edgeStateStyles,
                layout: layout,
            });
            if (graphRef) {
                graphRef.current = graph;
            }
        }
        processMinimap(minimapCfg, graph);
        graph.data(data);
        graph.render();
        // modify the node color according to the in edge
        var edges = graph.getEdges();
        edges.forEach(function (edge) {
            var line = edge.getKeyShape();
            var stroke = line.attr('stroke');
            var targetNode = edge.getTarget();
            targetNode.update({
                style: {
                    stroke: stroke,
                },
            });
        });
        graph.fitView();
        graph.on('edge:mouseenter', function (evt) {
            var item = evt.item;
            graph.setItemState(item, 'hover', true);
            if (handleEdgeHover) {
                handleEdgeHover(item, graph);
            }
        });
        graph.on('edge:mouseleave', function (evt) {
            var item = evt.item;
            graph.setItemState(item, 'hover', false);
            if (handleEdgeUnHover) {
                handleEdgeUnHover(item, graph);
            }
        });
        graph.on('edge:click', function (evt) {
            var item = evt.item;
            if (handleEdgeClick) {
                handleEdgeClick(item, graph);
            }
        });
        graph.on('node:mouseenter', function (evt) {
            var item = evt.item;
            graph.setItemState(item, 'hover', false);
            if (handleNodeHover) {
                handleNodeHover(item, graph);
            }
        });
        graph.on('node:mouseleave', function (evt) {
            var item = evt.item;
            graph.setItemState(item, 'hover', false);
            if (handleNodeUnHover) {
                handleNodeUnHover(item, graph);
            }
        });
        graph.on('node:click', function (evt) {
            var item = evt.item;
            if (handleNodeClick) {
                handleNodeClick(item, graph);
            }
        });
        graph.on('canvas:click', function (evt) {
            handleCanvasClick && handleCanvasClick(graph);
        });
        return function () { return graph.destroy(); };
    }, []);
    return (React.createElement(ErrorBoundary, null,
        React.createElement("div", { className: className, style: style, ref: container })));
};
export default DagreFundFlowGraph;
