// @antv/g2plot/lib/plots/funnel/geometries/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.conversionTagComponent = exports.geometryLabel = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var utils_1 = require("../../../utils");
var constant_1 = require("../constant");
/**
 * 漏斗图通用geometry label
 * @param geometry 对应的 chart geometry
 */
function geometryLabel(geometry) {
    return function (params) {
        var options = params.options;
        var xField = options.xField, yField = options.yField, label = options.label;
        if (!label) {
            geometry.label(false);
        }
        else {
            var callback = label.callback, cfg = tslib_1.__rest(label, ["callback"]);
            geometry.label({
                fields: [xField, yField, constant_1.FUNNEL_PERCENT, constant_1.FUNNEL_CONVERSATION],
                callback: callback,
                cfg: utils_1.transformLabel(cfg),
            });
        }
        return params;
    };
}
exports.geometryLabel = geometryLabel;
/**
 * 漏斗图通用转化率组件
 * @param getLineCoordinate 用于获取特定的 line 的位置及配置
 */
function conversionTagComponent(getLineCoordinate) {
    return function (params) {
        var chart = params.chart, options = params.options;
        var conversionTag = options.conversionTag;
        var data = chart.getOptions().data;
        if (conversionTag) {
            var formatter_1 = conversionTag.formatter;
            data.forEach(function (obj, index) {
                if (index <= 0)
                    return;
                var lineOption = getLineCoordinate(obj, index, data, {
                    top: true,
                    text: {
                        content: util_1.isFunction(formatter_1) ? formatter_1(obj, data) : formatter_1,
                        offsetX: conversionTag.offsetX,
                        offsetY: conversionTag.offsetY,
                        position: 'end',
                        autoRotate: false,
                        style: tslib_1.__assign({ textAlign: 'start', textBaseline: 'middle' }, conversionTag.style),
                    },
                });
                chart.annotation().line(lineOption);
            });
        }
        return params;
    };
}
exports.conversionTagComponent = conversionTagComponent;
