// @antv/g2plot/lib/plots/bullet/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bullet = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var utils_1 = require("../../utils");
var adaptor_1 = require("./adaptor");
var Bullet = /** @class */ (function (_super) {
    tslib_1.__extends(Bullet, _super);
    function Bullet() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'bullet';
        return _this;
    }
    /**
     * 获取子弹图的适配器
     */
    Bullet.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    Bullet.prototype.getDefaultOptions = function () {
        return utils_1.deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            layout: 'horizontal',
            size: {
                range: 30,
                measure: 20,
                target: 20,
            },
            xAxis: {
                tickLine: false,
            },
            bulletStyle: {
                range: {
                    fillOpacity: 0.5,
                },
            },
            label: {
                measure: {
                    position: 'right',
                },
            },
        });
    };
    return Bullet;
}(plot_1.Plot));
exports.Bullet = Bullet;
