// @antv/g2plot/esm/plots/bar/index.js
import { __extends } from "tslib";
import { Plot } from '../../core/plot';
import { deepAssign } from '../../utils';
import { getDataWhetherPecentage } from '../../utils/transform/percent';
import { adaptor } from './adaptor';
/**
 * 条形图
 */
var Bar = /** @class */ (function (_super) {
    __extends(Bar, _super);
    function Bar() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'bar';
        return _this;
    }
    /**
     * @override
     */
    Bar.prototype.changeData = function (data) {
        this.updateOption({ data: data });
        var _a = this.options, xField = _a.xField, yField = _a.yField, isPercent = _a.isPercent;
        this.chart.changeData(getDataWhetherPecentage(data, xField, yField, xField, isPercent));
    };
    /**
     * 获取 条形图 默认配置
     */
    Bar.prototype.getDefaultOptions = function () {
        return deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            barWidthRatio: 0.6,
            marginRatio: 1 / 32,
            tooltip: {
                shared: true,
                showMarkers: false,
                offset: 20,
            },
            interactions: [{ type: 'active-region' }],
        });
    };
    /**
     * 获取 条形图 的适配器
     */
    Bar.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    return Bar;
}(Plot));
export { Bar };
