// @antv/g2plot/lib/plots/funnel/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = exports.meta = void 0;
var common_1 = require("../../adaptor/common");
var utils_1 = require("../../utils");
var basic_1 = require("./geometries/basic");
var compare_1 = require("./geometries/compare");
var dynamic_height_1 = require("./geometries/dynamic-height");
var constant_1 = require("./constant");
/**
 *
 * 各式漏斗图geometry实现细节有较大不同,
 * 1. 普通漏斗图：interval.shape('funnel')
 * 2. 对比漏斗图：分面
 * 3. 动态高度漏斗图：polypon
* /

/**
 * options 处理
 * @param params
 */
function defaultOptions(params) {
    var options = params.options;
    var compareField = options.compareField, xField = options.xField, yField = options.yField;
    var defaultOption = {
        label: {
            offset: 0,
            position: 'middle',
            style: {
                fill: '#fff',
                fontSize: 12,
            },
            formatter: compareField ? function (datum) { return "" + datum[yField]; } : function (datum) { return datum[xField] + " " + datum[yField]; },
        },
        tooltip: {
            showTitle: false,
            showMarkers: false,
            shared: false,
            title: xField,
            formatter: function (datum) {
                return { name: datum[xField], value: datum[yField] };
            },
        },
        conversionTag: {
            offsetX: 10,
            offsetY: 0,
            style: {},
            formatter: function (datum) { return "\u8F6C\u5316\u7387" + (datum[constant_1.FUNNEL_CONVERSATION] * 100).toFixed(2) + "%"; },
        },
    };
    return utils_1.deepAssign({
        options: defaultOption,
    }, params);
}
/**
 * geometry处理
 * @param params
 */
function geometry(params) {
    var options = params.options;
    var compareField = options.compareField, dynamicHeight = options.dynamicHeight;
    if (compareField) {
        return compare_1.compareFunnel(params);
    }
    if (dynamicHeight) {
        return dynamic_height_1.dynamicHeightFunnel(params);
    }
    return basic_1.basicFunnel(params);
}
/**
 * meta 配置
 * @param params
 */
function meta(params) {
    var _a;
    var options = params.options;
    var xAxis = options.xAxis, yAxis = options.yAxis, xField = options.xField, yField = options.yField;
    return utils_1.flow(common_1.scale((_a = {},
        _a[xField] = xAxis,
        _a[yField] = yAxis,
        _a)))(params);
}
exports.meta = meta;
/**
 * 坐标轴
 * @param params
 */
function axis(params) {
    var chart = params.chart;
    chart.axis(false);
    return params;
}
/**
 * legend 配置
 * @param params
 */
function legend(params) {
    var chart = params.chart, options = params.options;
    var legend = options.legend;
    if (legend === false) {
        chart.legend(false);
    }
    else {
        chart.legend(legend);
        // TODO FIX: legend-click 时间和转化率组件之间的关联
    }
    return params;
}
/**
 * 漏斗图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    return utils_1.flow(defaultOptions, geometry, meta, axis, common_1.tooltip, common_1.interaction, legend, common_1.animation, common_1.theme, common_1.annotation())(params);
}
exports.adaptor = adaptor;
