// @antv/g2plot/esm/plots/liquid/index.js
import { __extends } from "tslib";
import { Plot } from '../../core/plot';
import { adaptor } from './adaptor';
// register liquid shape
import './shapes/liquid';
/**
 * 传说中的水波图
 */
var Liquid = /** @class */ (function (_super) {
    __extends(Liquid, _super);
    function Liquid() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'liquid';
        return _this;
    }
    Liquid.prototype.getDefaultOptions = function () {
        return {
            color: '#6a99f9',
            radius: 0.9,
            statistic: {
                title: false,
                content: {
                    formatter: function (_a) {
                        var percent = _a.percent;
                        return (percent * 100).toFixed(2) + "%";
                    },
                    style: {
                        opacity: 0.75,
                        fontSize: '30px',
                        textAlign: 'center',
                    },
                },
            },
        };
    };
    /**
     * 更新数据
     * @param percent
     */
    Liquid.prototype.changeData = function (percent) {
        this.update({
            percent: percent,
        });
    };
    /**
     * 获取适配器
     */
    Liquid.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    return Liquid;
}(Plot));
export { Liquid };
