// @antv/g6/lib/plugins/menu/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _tslib = require("tslib");

var _modifyCss = _interopRequireDefault(require("@antv/dom-util/lib/modify-css"));

var _createDom = _interopRequireDefault(require("@antv/dom-util/lib/create-dom"));

var _isString = _interopRequireDefault(require("@antv/util/lib/is-string"));

var _insertCss = _interopRequireDefault(require("insert-css"));

var _base = _interopRequireDefault(require("../base"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _insertCss.default)("\n  .g6-component-contextmenu {\n    border: 1px solid #e2e2e2;\n    border-radius: 4px;\n    font-size: 12px;\n    color: #545454;\n    background-color: rgba(255, 255, 255, 0.9);\n    padding: 10px 8px;\n    box-shadow: rgb(174, 174, 174) 0px 0px 10px;\n  }\n  .g6-contextmenu-ul {\n    padding: 0;\n    margin: 0;\n    list-style: none;\n  }\n\n");

var Menu =
/** @class */
function (_super) {
  (0, _tslib.__extends)(Menu, _super);

  function Menu() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  Menu.prototype.getDefaultCfgs = function () {
    return {
      offsetX: 6,
      offsetY: 6,
      handleMenuClick: undefined,
      // 指定菜单内容，function(e) {...}
      getContent: function getContent(e) {
        return "\n          <ul class='g6-contextmenu-ul'>\n            <li>\u83DC\u5355\u98791</li>\n            <li>\u83DC\u5355\u98792</li>\n          </ul>\n        ";
      },
      shouldBegin: function shouldBegin(e) {
        return true;
      },
      // 菜单隐藏事件
      onHide: function onHide() {
        return true;
      },
      itemTypes: ['node', 'edge', 'combo']
    };
  }; // class-methods-use-this


  Menu.prototype.getEvents = function () {
    return {
      contextmenu: 'onMenuShow'
    };
  };

  Menu.prototype.init = function () {
    var className = this.get('className');
    var menu = (0, _createDom.default)("<div class=" + (className || 'g6-component-contextmenu') + "></div>");
    (0, _modifyCss.default)(menu, {
      top: '0px',
      position: 'absolute',
      visibility: 'hidden'
    });
    var container = this.get('container');

    if (!container) {
      container = this.get('graph').get('container');
    }

    container.appendChild(menu);
    this.set('menu', menu);
  };

  Menu.prototype.onMenuShow = function (e) {
    var self = this;
    e.preventDefault();
    e.stopPropagation();
    if (!e.item) return;
    var shouldBegin = this.get('shouldBegin');
    if (!shouldBegin(e)) return;
    var itemTypes = this.get('itemTypes');

    if (e.item && e.item.getType && itemTypes.indexOf(e.item.getType()) === -1) {
      self.onMenuHide();
      return;
    }

    if (!e.item) {
      return;
    }

    var menuDom = this.get('menu');
    var getContent = this.get('getContent');
    var graph = this.get('graph');
    var menu = getContent(e, graph);

    if ((0, _isString.default)(menu)) {
      menuDom.innerHTML = menu;
    } else {
      menuDom.innerHTML = menu.outerHTML;
    } // 清除之前监听的事件


    this.removeMenuEventListener();
    var handleMenuClick = this.get('handleMenuClick');

    if (handleMenuClick) {
      var handleMenuClickWrapper = function handleMenuClickWrapper(evt) {
        handleMenuClick(evt.target, e.item, graph);
      };

      this.set('handleMenuClickWrapper', handleMenuClickWrapper);
      menuDom.addEventListener('click', handleMenuClickWrapper);
    }

    var width = graph.get('width');
    var height = graph.get('height');
    var bbox = menuDom.getBoundingClientRect();
    var offsetX = this.get('offsetX') || 0;
    var offsetY = this.get('offsetY') || 0;
    var graphTop = graph.getContainer().offsetTop;
    var graphLeft = graph.getContainer().offsetLeft;
    var x = e.canvasX + graphLeft + offsetX;
    var y = e.canvasY + graphTop + offsetY; // when the menu is (part of) out of the canvas

    if (x + bbox.width > width) {
      x = e.canvasX - bbox.width - offsetX + graphLeft;
    }

    if (y + bbox.height > height) {
      y = e.canvasY - bbox.height - offsetY + graphTop;
    }

    (0, _modifyCss.default)(menuDom, {
      top: y + "px",
      left: x + "px",
      visibility: 'visible'
    });

    var handler = function handler(evt) {
      self.onMenuHide();
    }; // 如果在页面中其他任意地方进行click, 隐去菜单


    document.body.addEventListener('click', handler);
    this.set('handler', handler);
  };

  Menu.prototype.removeMenuEventListener = function () {
    var handleMenuClickWrapper = this.get('handleMenuClickWrapper');
    var handler = this.get('handler');

    if (handleMenuClickWrapper) {
      var menuDom = this.get('menu');
      menuDom.removeEventListener('click', handleMenuClickWrapper);
      this.set('handleMenuClickWrapper', null);
    }

    if (handler) {
      document.body.removeEventListener('click', handler);
    }
  };

  Menu.prototype.onMenuHide = function () {
    var menuDom = this.get('menu');

    if (menuDom) {
      (0, _modifyCss.default)(menuDom, {
        visibility: 'hidden'
      });
    } // 隐藏菜单后需要移除事件监听


    this.removeMenuEventListener();
  };

  Menu.prototype.destroy = function () {
    var menu = this.get('menu');
    this.removeMenuEventListener();

    if (menu) {
      var container = this.get('container');

      if (!container) {
        container = this.get('graph').get('container');
      }

      container.removeChild(menu);
    }
  };

  return Menu;
}(_base.default);

var _default = Menu;
exports.default = _default;
