// @antv/g2plot/esm/plots/tiny-area/adaptor.js
import { theme, scale, animation, annotation, tooltip } from '../../adaptor/common';
import { flow, deepAssign } from '../../utils';
import { area, line, point } from '../../adaptor/geometries';
import { X_FIELD, Y_FIELD } from '../tiny-line/constants';
import { adjustYMetaByZero } from '../../utils/data';
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var _a, _b;
    var chart = params.chart, options = params.options;
    var data = options.data, xAxis = options.xAxis, yAxis = options.yAxis, color = options.color, areaStyle = options.areaStyle, pointOptions = options.point, lineOptions = options.line;
    var seriesData = data.map(function (y, x) {
        return { x: x, y: y };
    });
    chart.data(seriesData);
    var p = deepAssign({}, params, {
        options: {
            xField: X_FIELD,
            yField: Y_FIELD,
            area: { color: color, style: areaStyle },
            line: lineOptions,
            point: pointOptions,
        },
    });
    // area geometry 处理
    area(p);
    line(p);
    point(p);
    chart.axis(false);
    chart.legend(false);
    // scale
    scale((_a = {},
        _a[X_FIELD] = xAxis,
        _a[Y_FIELD] = yAxis,
        _a), (_b = {},
        _b[X_FIELD] = {
            type: 'cat',
        },
        _b[Y_FIELD] = adjustYMetaByZero(seriesData, Y_FIELD),
        _b))(params);
    return params;
}
/**
 * 迷你面积图适配器
 * @param chart
 * @param options
 */
export function adaptor(params) {
    return flow(geometry, tooltip, theme, animation, annotation())(params);
}
