// @antv/g-base/lib/index.js
"use strict";
/**
 * @fileoverview G 的基础接口定义和所有的抽象类
 * @author dxq613@gmail.com
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathUtil = exports.version = void 0;
var tslib_1 = require("tslib");
var PathUtil = require("./util/path");
exports.PathUtil = PathUtil;
var pkg = require('../package.json');
exports.version = pkg.version;
tslib_1.__exportStar(require("./types"), exports);
tslib_1.__exportStar(require("./interfaces"), exports);
var graph_event_1 = require("./event/graph-event");
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return graph_event_1.default; } });
var base_1 = require("./abstract/base");
Object.defineProperty(exports, "Base", { enumerable: true, get: function () { return base_1.default; } });
var canvas_1 = require("./abstract/canvas");
Object.defineProperty(exports, "AbstractCanvas", { enumerable: true, get: function () { return canvas_1.default; } });
var group_1 = require("./abstract/group");
Object.defineProperty(exports, "AbstractGroup", { enumerable: true, get: function () { return group_1.default; } });
var shape_1 = require("./abstract/shape");
Object.defineProperty(exports, "AbstractShape", { enumerable: true, get: function () { return shape_1.default; } });
