// @antv/g2plot/esm/plots/tiny-column/adaptor.js
import { theme, scale, animation, annotation, tooltip } from '../../adaptor/common';
import { flow, deepAssign } from '../../utils';
import { interval } from '../../adaptor/geometries';
import { X_FIELD, Y_FIELD } from '../tiny-line/constants';
import { getTinyData } from '../tiny-line/utils';
import { adjustYMetaByZero } from '../../utils/data';
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var _a, _b;
    var chart = params.chart, options = params.options;
    var data = options.data, xAxis = options.xAxis, yAxis = options.yAxis, color = options.color, columnStyle = options.columnStyle, columnWidthRatio = options.columnWidthRatio;
    var seriesData = getTinyData(data);
    chart.data(seriesData);
    var p = deepAssign({}, params, {
        options: {
            xField: X_FIELD,
            yField: Y_FIELD,
            widthRatio: columnWidthRatio,
            interval: {
                style: columnStyle,
                color: color,
            },
        },
    });
    interval(p);
    chart.axis(false);
    chart.legend(false);
    chart.interaction('element-active');
    // scale
    scale((_a = {},
        _a[X_FIELD] = xAxis,
        _a[Y_FIELD] = yAxis,
        _a), (_b = {},
        _b[X_FIELD] = {
            type: 'cat',
        },
        _b[Y_FIELD] = adjustYMetaByZero(seriesData, Y_FIELD),
        _b))(params);
    return params;
}
/**
 * 迷你柱形图适配器
 * @param chart
 * @param options
 */
export function adaptor(params) {
    return flow(geometry, tooltip, theme, animation, annotation())(params);
}
