// @antv/g2plot/esm/plots/tiny-column/index.js
import { __assign, __extends } from "tslib";
import { Plot } from '../../core/plot';
import { getTinyData } from '../tiny-line/utils';
import { adaptor } from './adaptor';
import { DEFAULT_TOOLTIP_OPTIONS } from './constants';
var TinyColumn = /** @class */ (function (_super) {
    __extends(TinyColumn, _super);
    function TinyColumn() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'tiny-column';
        return _this;
    }
    /**
     * @override
     * @param data
     */
    TinyColumn.prototype.changeData = function (data) {
        this.updateOption({ data: data });
        this.chart.changeData(getTinyData(data));
    };
    TinyColumn.prototype.getDefaultOptions = function () {
        return {
            appendPadding: 2,
            tooltip: __assign({}, DEFAULT_TOOLTIP_OPTIONS),
            animation: true,
        };
    };
    /**
     * 获取 迷你柱形图 的适配器
     */
    TinyColumn.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    return TinyColumn;
}(Plot));
export { TinyColumn };
