// @antv/g2plot/esm/plots/progress/index.js
import { __extends } from "tslib";
import { Plot } from '../../core/plot';
import { adaptor } from './adaptor';
import { DEFAULT_COLOR } from './constant';
import { getProgressData } from './utils';
var Progress = /** @class */ (function (_super) {
    __extends(Progress, _super);
    function Progress() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'process';
        return _this;
    }
    Progress.prototype.getDefaultOptions = function () {
        return {
            percent: 0.2,
            color: DEFAULT_COLOR,
            animation: true,
        };
    };
    /**
     * 更新数据
     * @param percent
     */
    Progress.prototype.changeData = function (percent) {
        this.updateOption({ percent: percent });
        this.chart.changeData(getProgressData(percent));
    };
    /**
     * 获取 进度图 的适配器
     */
    Progress.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    return Progress;
}(Plot));
export { Progress };
