// @antv/g2plot/esm/plots/area/index.js
import { __extends } from "tslib";
import { Plot } from '../../core/plot';
import { deepAssign } from '../../utils';
import { getDataWhetherPecentage } from '../../utils/transform/percent';
import { adaptor } from './adaptor';
var Area = /** @class */ (function (_super) {
    __extends(Area, _super);
    function Area() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'area';
        return _this;
    }
    /**
     * 获取 折线图 默认配置
     */
    Area.prototype.getDefaultOptions = function () {
        return deepAssign({}, _super.prototype.getDefaultOptions.call(this), {
            tooltip: {
                shared: true,
                showMarkers: true,
                showCrosshairs: true,
                crosshairs: {
                    type: 'x',
                },
            },
            isStack: true,
            // 默认开启
            line: {},
            legend: {
                position: 'top-left',
            },
        });
    };
    /**
     * @override
     * @param data
     */
    Area.prototype.changeData = function (data) {
        this.updateOption({ data: data });
        var _a = this.options, isPercent = _a.isPercent, xField = _a.xField, yField = _a.yField;
        this.chart.changeData(getDataWhetherPecentage(data, yField, xField, yField, isPercent));
    };
    /**
     * 获取 面积图 的适配器
     */
    Area.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    return Area;
}(Plot));
export { Area };
