// @antv/g2plot/esm/plots/treemap/adaptor.js
import { polygon as basePolygon } from '../../adaptor/geometries/polygon';
import { interaction, animation, theme, legend, annotation, tooltip } from '../../adaptor/common';
import { flow, deepAssign } from '../../utils';
import { transformData } from './utils';
/**
 * 获取默认 option
 * @param params
 */
function defaultOptions(params) {
    var options = params.options;
    var colorField = options.colorField;
    return deepAssign({
        options: {
            // 默认按照 name 字段对颜色进行分类
            colorField: 'name',
            label: {
                fields: ['name'],
                layout: {
                    type: 'limit-in-shape',
                },
            },
            tooltip: {
                showMarkers: false,
                showTitle: false,
                fields: ['name', 'value', colorField],
                formatter: function (data) {
                    return {
                        name: data.name,
                        value: data.value,
                    };
                },
            },
        },
    }, params);
}
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var color = options.color, colorField = options.colorField;
    var data = transformData(options);
    chart.data(data);
    // geometry
    basePolygon(deepAssign({}, params, {
        options: {
            xField: 'x',
            yField: 'y',
            seriesField: colorField,
            polygon: {
                color: color,
                style: {
                    lineWidth: 1,
                    stroke: '#fff',
                },
            },
        },
    }));
    return params;
}
/**
 * 坐标轴
 * @param params
 */
function axis(params) {
    var chart = params.chart;
    chart.axis(false);
    return params;
}
/**
 * 矩形树图
 * @param chart
 * @param options
 */
export function adaptor(params) {
    return flow(defaultOptions, geometry, axis, theme, legend, tooltip, interaction, animation, annotation())(params);
}
