// @antv/g2plot/lib/plots/treemap/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformData = void 0;
var treemap_1 = require("../../utils/hierarchy/treemap");
function transformData(options) {
    var data = options.data, colorField = options.colorField;
    var nodes = treemap_1.treemap(data, {
        // @ts-ignore
        type: 'hierarchy.treemap',
        tile: 'treemapResquarify',
        field: 'value',
        as: ['x', 'y'],
    });
    var result = [];
    nodes.forEach(function (node) {
        if (!node.children) {
            var eachNode = Object.assign({}, node.data, {
                x: node.x,
                y: node.y,
                depth: node.depth,
                value: node.value,
            });
            if (!node.data[colorField] && node.parent) {
                var ancestorNode = node.ancestors().find(function (n) { return n.data[colorField]; });
                eachNode[colorField] = ancestorNode === null || ancestorNode === void 0 ? void 0 : ancestorNode.data[colorField];
            }
            else {
                eachNode[colorField] = node.data[colorField];
            }
            result.push(eachNode);
        }
    });
    return result;
}
exports.transformData = transformData;
