// @antv/g2plot/lib/plots/radial-bar/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = exports.axis = exports.meta = void 0;
var common_1 = require("../../adaptor/common");
var utils_1 = require("../../utils");
var geometries_1 = require("../../adaptor/geometries");
var utils_2 = require("./utils");
/**
 * geometry 处理
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var data = options.data, style = options.barStyle, color = options.color, tooltip = options.tooltip, colorField = options.colorField, type = options.type, xField = options.xField, yField = options.yField;
    chart.data(data);
    var p = utils_1.deepAssign({}, params, {
        options: {
            tooltip: tooltip,
            seriesField: colorField,
            interval: {
                style: style,
                color: color,
                shape: type === 'line' ? 'line' : 'intervel',
            },
        },
    });
    geometries_1.interval(p);
    if (type === 'line') {
        chart.point().position(xField + "*" + yField).shape('circle');
    }
    return params;
}
/**
 * meta 配置
 * @param params
 */
function meta(params) {
    var _a;
    var options = params.options;
    var yField = options.yField, data = options.data, maxAngle = options.maxAngle;
    return utils_1.flow(common_1.scale((_a = {},
        _a[yField] = {
            min: 0,
            max: utils_2.getScaleMax(maxAngle, yField, data),
        },
        _a)))(params);
}
exports.meta = meta;
/**
 * coordinate 配置
 * @param params
 */
function coordinate(params) {
    var chart = params.chart, options = params.options;
    var radius = options.radius, innerRadius = options.innerRadius;
    chart
        .coordinate({
        type: 'polar',
        cfg: {
            radius: radius,
            innerRadius: innerRadius,
        },
    })
        .transpose();
    return params;
}
/**
 * axis 配置
 * @param params
 */
function axis(params) {
    var chart = params.chart, options = params.options;
    var xField = options.xField, xAxis = options.xAxis;
    chart.axis(xField, xAxis);
    return params;
}
exports.axis = axis;
/**
 * 图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    return utils_1.flow(geometry, meta, axis, coordinate, common_1.interaction, common_1.animation, common_1.theme, common_1.tooltip, common_1.legend, common_1.annotation())(params);
}
exports.adaptor = adaptor;
