// @antv/g2plot/lib/plots/progress/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Progress = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
var constant_1 = require("./constant");
var Progress = /** @class */ (function (_super) {
    tslib_1.__extends(Progress, _super);
    function Progress() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'process';
        return _this;
    }
    Progress.prototype.getDefaultOptions = function () {
        return {
            percent: 0.2,
            color: constant_1.DEFAULT_COLOR,
        };
    };
    /**
     * 更新数据
     * @param percent
     */
    Progress.prototype.changeData = function (percent) {
        this.update({
            percent: percent,
        });
    };
    /**
     * 获取 进度图 的适配器
     */
    Progress.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Progress;
}(plot_1.Plot));
exports.Progress = Progress;
